src/rendering/mod.rs

```diff
@@ -176,8 +178,12 @@ impl GlState {
         gl.ActiveTexture(ffi::TEXTURE0);
         gl.BindVertexArray(self.vertex_array);
 
+        let mut damage = Vec::from(damage);
+
+        damage.dedup();
```

Comment:
Deduplication is not really enough, we want to merge overlapping rectangles.

You can just steal these lines from smithay to do that:
https://github.com/Smithay/smithay/blob/master/src/desktop/space/mod.rs#L508-L519

Also we could pre-filter the damage to overlap with the screen-space, by scaling it via `to_logical` and filtering by `overlaps`. Then we don't have to do that in `paint_mesh`.